rfcs/0003-declaration-engine.md

```diff
@@ -0,0 +1,189 @@
+
+For example, imagine that our toy language includes variables and variable assignments. If we have a variable `x` of type `bool`, and then declare a different variable `y` and assign it to `x`, we can infer that `y` is of type `bool`.
+
+But, this property is not exclusive to just `bool`s---variable assignment does not "work differently" for `bool`s than it does for `u64`s. Thus, we can make this example more abstract. If `x` is of type `T`, and we assign `y` to `x`, we can infer that `y` is of type `T`. Importantly, `T` could be any concrete type (`bool`, `u64`, etc), and that would not affect the type inference that we just made---no matter what `T` is, we still know that `y` is of type `T` because `x` is of type `T`.
+
+But we take this one step furthur... when performing type inference, we don't even really need to know that `y` is type `T`, we simply only need to know that `y` _has the same type as `x`_. This is a constraint. The toy compiler can perform type inference for `y` by generating the constraint "`y` has the same type as `x`".
+
+Now, imagine that the toy language includes if statements and the toy compiler encounters an if statement that uses `x` as the conditional. The toy compiler has this list of constraints:
+
+1. `x` has type `T`
+2. `y` has the same type as `x`
+
+But when `x` is used as the conditional in an if statement, we can add an additional constraint:
+
+3. the type of `x` is `bool`
+
+With these three constraints, the toy compiler is able to perform constraint solving (unifying the constraints with unification) to determine that both `x` and `y` are of type `bool`.
+
+Now, imagine that the toy language includes addition expressions and the toy compiler encounters an addition expression `y + 1u64`. This would generate the constraint:
+
+4. `y` has type `u64`
+
+With these four constraints, the toy compiler is unable to perform unification, as the constraints are unable to be unified and generate a type error.
+
+We can see that the toy compiler would be able to perform type inference for a whole program by generating a list of constraints and then unifying these constraints.
+
+(Type inference, constraint solving, and unification are defined here: https://papl.cs.brown.edu/2014/Type_Inference.html)
+
+## Declaration Engine
+
+The declaration engine takes the concept of constraints and unification and applies it to declarations.
+
+For example, when type checking a function application, the compiler does not need to care about the contents of the body of the function, instead it is sufficient to create a constraint referring to "the return type of the function".
+
+For example, given a typed function declaration `add`, we add this declaration to the declaration engine. Then when type checking a function application of `add`, instead of inlining the function body, we create constraints referring to "the body of `add`" and "the return type of `add`". When type checking a _function application_
```

Comment:
```suggestion
For example, given a typed function declaration `add`, we add this declaration to the declaration engine. Then when type checking a function application of `add`, instead of inlining the function body, we create constraints referring to "the body of `add`" and "the return type of `add`". When type checking a _function application_.
```

Reply:
Would the type of the body of the function not be the same as the return type of the function? I would think that the only constraints that would matter would be those of the function inputs, the type parameters, and then that of the function return type.